src/Repository/AbstractRepository.php

```
@@ -0,0 +1,290 @@
+<?php
+
+namespace Minormous\Dali\Repository;
+
+use Auryn\Injector;
+use Doctrine\DBAL\Query\QueryBuilder;
+use Generator;
+use Minormous\Metabolize\Dali\Metadata;
+use Minormous\Dali\Driver\AbstractDriver;
+use Minormous\Dali\Driver\Interfaces\QueryResultInterface;
+use Minormous\Dali\Entity\Interfaces\CastInterface;
+use Minormous\Dali\Entity\Interfaces\EntityInterface;
+use Minormous\Dali\Exceptions\EntityDoesNotExistException;
+use Minormous\Dali\Exceptions\InvalidEntityException;
+use Minormous\Dali\Exceptions\InvalidRepositoryException;
+use ReflectionClass;
+use Traversable;
+use Webmozart\Assert\Assert;
+
+abstract class AbstractRepository
+{
+    final public function __construct(
+        protected AbstractDriver $driver,
+        protected Metadata $metadata,
+        protected Injector $container,
+    ) {
+    }
+
+    public function getDriver(): AbstractDriver
+    {
+        return $this->driver;
+    }
+
+    public function findById(string|int $id): ?EntityInterface
+    {
+        $idMetadata = $this->metadata->getIdColumn();
+        if (!$idMetadata) {
+            throw InvalidEntityException::noIdColumn($this->metadata->getClass());
+        }
+        [$key, $value] = $this->convertValueForDriver($idMetadata->getProperty(), $id);
+        $value = $this->driver->findOne($this->metadata->getTable(), [$key => $value]);
```

again, need to allow for multi-field identifiers

---

Also should we change this to throw an exception if we do not find a record instead of just returning null?